### PR TITLE
refactor(BaseConfig): update docstring, extract factory method and remove unnecessary variable assignment

### PR DESCRIPTION
**Pull request by bearomorphism**
## Checklist

- [ ] I have read the [contributing guidelines](https://commitizen-tools.github.io/commitizen/contributing/)

### Code Changes

- [ ] Add test cases to all the changes you introduce
- [ ] Run `poetry all` locally to ensure this change passes linter check and tests
- [ ] Manually test the changes:
  - [ ] Verify the feature/bug fix works as expected in real-world scenarios
  - [ ] Test edge cases and error conditions
  - [ ] Ensure backward compatibility is maintained
  - [ ] Document any manual testing steps performed
- [ ] Update the documentation for the changes

### Documentation Changes

- [ ] Run `poetry doc` locally to ensure the documentation pages renders correctly
- [ ] Check and fix any broken links (internal or external) in the documentation

> When running `poetry doc`, any broken internal documentation links will be reported in the console output like this:
>
> ```text
> INFO    -  Doc file 'config.md' contains a link 'commands/bump.md#-post_bump_hooks', but the doc 'commands/bump.md' does not contain an anchor '#-post_bump_hooks'.
> ```

### CODE REVIEW

File: commitizen/config/json_config.py [hunk 1]
```diff
@@ -22,7 +22,6 @@
 class JsonConfig(BaseConfig):
     def __init__(self, *, data: bytes | str, path: Path | str) -> None:
```
Comment (bearomorphism) [inline commitizen/config/json_config.py]: I think we can remove the `str` of the parameter `path` in the future? I'm not sure if that will be a breaking change

File: commitizen/config/base_config.py [hunk 2]
```diff
@@ -17,8 +17,8 @@
 
 class BaseConfig:
     def __init__(self) -> None:
+        self.is_empty_config = False
         self._settings: Settings = DEFAULT_SETTINGS.copy()
-        self.encoding = self.settings["encoding"]
```
Comment (bearomorphism) [inline commitizen/config/base_config.py:21]: I'm not sure why this member variable exists. It will be stale if `self._settings` is updated somewhere.

File: commitizen/config/base_config.py [hunk 3]
```diff
@@ -17,8 +17,8 @@
 
 class BaseConfig:
```
Comment (bearomorphism) [inline commitizen/config/base_config.py:18]: Should this be an abstract class?

Comment (woile) [reply commitizen/config/base_config.py:18]: Ideally, a `Protocol`, but it doesn't hurt like this

File: commitizen/config/json_config.py (hunk 1 above)
Comment (woile) [reply commitizen/config/json_config.py]: I think it's fine to remove the `str`, at most, it would generate a `mypy` error

File: commitizen/config/base_config.py (hunk 3 above)
Comment (bearomorphism) [reply commitizen/config/base_config.py:18]: Thanks. I found it odd that we can create `BaseConfig` objects and it's sometimes confusing. I attempted to refactor it on another branch (not published yet) but many tests would be affected.

File: commitizen/config/json_config.py (hunk 1 above)
Comment (bearomorphism) [reply commitizen/config/json_config.py]: removed `| str`

File: commitizen/config/base_config.py (hunk 3 above)
Comment (Lee-W) [reply commitizen/config/base_config.py:18]: Let's keep it as it is now and improve later

**Review by Lee-W — APPROVED** [review]
